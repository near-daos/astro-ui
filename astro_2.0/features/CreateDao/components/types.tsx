import { IconName } from 'components/Icon';

export type DAOType = 'club' | 'cooperative' | 'corporation' | 'foundation';
export type Subject = 'proposals' | 'structure' | 'voting';
export type DAOVotingPowerType = 'democratic' | 'weighted';
export type DAOProposalsType = 'open' | 'closed';
export type DAOStructureType = 'flat' | 'groups';
export type DaoImageType = 'flagCover' | 'flagLogo' | 'tokenImage';

export interface DAOTemplate {
  title: string;
  description: string;
  variant: DAOType;
  proposals: DAOProposalsType;
  structure: DAOStructureType;
  voting: DAOVotingPowerType;
  disabled?: boolean;
}

export type DaoSettingOption<T> = {
  title: string;
  value: T;
  icon: IconName;
  subject: Subject;
  description: string;
  disabled?: boolean;
};

export interface IDaoCreateForm {
  address: string;
  displayName: string;
  purpose: string;
  websites: string[];
  flagCover: FileList;
  flagLogo: FileList;
  defaultFlag: string;
}

export interface DAOFormValues extends IDaoCreateForm {
  proposals: DAOProposalsType;
  structure: DAOStructureType;
  voting: DAOVotingPowerType;
  legalStatus?: string;
  legalLink?: string;
  gas: number;
  members: string[];
}

export type DaoSubjectInfo = {
  subject: Subject;
  title: string;
  subTitle: string;
};
