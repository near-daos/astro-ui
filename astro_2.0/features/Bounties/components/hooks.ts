import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
