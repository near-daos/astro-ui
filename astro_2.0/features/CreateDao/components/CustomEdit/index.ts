import CustomEdit from './CustomEdit';

export default CustomEdit;
