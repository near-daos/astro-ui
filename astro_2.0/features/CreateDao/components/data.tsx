import {
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
  DaoSubjectInfo,
} from 'astro_2.0/features/CreateDao/components/types';

const DAO_PROPOSALS_OPEN: DaoSettingOption<DAOProposalsType> = {
  value: 'open',
  icon: 'illustrationOpenOrganization',
  subject: 'proposals',
  title: 'daoRulesAll',
  description: 'daoRulesAllDescription',
};

const DAO_PROPOSALS_CLOSED: DaoSettingOption<DAOProposalsType> = {
  value: 'closed',
  icon: 'illustrationMembersOnly',
  subject: 'proposals',
  title: 'daoRulesMembers',
  description: 'daoRulesMembersDescription',
};

const DAO_VOTING_POWER_DEMOCRATIC: DaoSettingOption<DAOVotingPowerType> = {
  value: 'democratic',
  icon: 'illustrationVotePerMember',
  subject: 'voting',
  title: 'daoRulesDemocratic',
  description: 'daoRulesDemocraticDescription',
};

const DAO_VOTING_POWER_TOKEN: DaoSettingOption<DAOVotingPowerType> = {
  value: 'weighted',
  icon: 'illustrationTokenWeighted',
  subject: 'voting',
  title: 'daoRulesTokenWeighted',
  description: 'daoRulesTokenWeightedDescription',
};

const DAO_STRUCTURE_FLAT: DaoSettingOption<DAOStructureType> = {
  value: 'flat',
  icon: 'illustrationFlatOrganization',
  subject: 'structure',
  title: 'daoRulesFlat',
  description: 'daoRulesFlatDescription',
};

const DAO_STRUCTURE_GROUPS: DaoSettingOption<DAOStructureType> = {
  value: 'groups',
  icon: 'illustrationGroupsAndCommitttees',
  subject: 'structure',
  title: 'daoRulesGroups',
  description: 'daoRulesGroupsDescription',
};

export const DAO_PROPOSALS_OPTIONS: Record<
  DAOProposalsType,
  DaoSettingOption<DAOProposalsType>
> = {
  open: DAO_PROPOSALS_OPEN,
  closed: DAO_PROPOSALS_CLOSED,
};

export const DAO_VOTING_POWER_OPTIONS: Record<
  DAOVotingPowerType,
  DaoSettingOption<DAOVotingPowerType>
> = {
  democratic: DAO_VOTING_POWER_DEMOCRATIC,
  weighted: DAO_VOTING_POWER_TOKEN,
};

export const DAO_STRUCTURE_OPTIONS: Record<
  DAOStructureType,
  DaoSettingOption<DAOStructureType>
> = {
  flat: DAO_STRUCTURE_FLAT,
  groups: DAO_STRUCTURE_GROUPS,
};

export const DAO_SUBJECT_OPTIONS = {
  voting: Object.values(DAO_VOTING_POWER_OPTIONS),
  structure: Object.values(DAO_STRUCTURE_OPTIONS),
  proposals: Object.values(DAO_PROPOSALS_OPTIONS),
};

export const DAO_RULES_INFO: DaoSubjectInfo[] = [
  {
    subject: 'proposals',
    title: 'daoRulesProposals',
    subTitle: 'daoRulesProposalsDescription',
  },
  {
    subject: 'voting',
    title: 'daoRulesVoting',
    subTitle: 'daoRulesVotingDescription',
  },
  {
    subject: 'structure',
    title: 'daoRulesStructure',
    subTitle: 'daoRulesStructureDescription',
  },
];
