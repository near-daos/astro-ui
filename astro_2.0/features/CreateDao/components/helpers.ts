import { GlobalState } from 'little-state-machine';
import {
  AssetsStep,
  GroupsStep,
  InfoStep,
  LegalStep,
  LinksStep,
  MembersStep,
  ProposalsStep,
  SubmitStep,
  VotingStep,
} from 'astro_2.0/features/CreateDao/types';
import { AnySchema } from 'yup';
import { ResolverError, ResolverResult } from 'react-hook-form';
import { DEFAULT_CREATE_DAO_GAS } from 'services/sputnik/constants';

export function getInitialValues(
  accountId: string,
  defaultFlag: string
): GlobalState {
  return {
    info: {
      displayName: '',
      address: '',
      purpose: '',
      isValid: false,
    },
    kyc: {
      legalStatus: '',
      legalLink: '',
      isValid: false,
    },
    links: {
      websites: [],
      isValid: false,
    },
    groups: {
      items: [{ name: 'council', slug: 'council' }],
      isValid: false,
    },
    members: {
      accounts: [],
      isValid: false,
    },
    assets: {
      flagCover: '',
      flagLogo: '',
      defaultFlag,
      isValid: false,
    },
    voting: {
      data: null,
      isValid: false,
    },
    proposals: {
      data: null,
      isValid: false,
    },
    submit: {
      isValid: true,
      gas: DEFAULT_CREATE_DAO_GAS,
    },
  };
}

export function updateAction(
  state: GlobalState,
  payload: Record<
    string,
    | InfoStep
    | LegalStep
    | LinksStep
    | MembersStep
    | GroupsStep
    | AssetsStep
    | VotingStep
    | ProposalsStep
    | SubmitStep
  >
): GlobalState {
  return {
    ...state,
    ...payload,
  };
}

export async function handleValidate<T>(
  schema: AnySchema,
  data: T,
  onUpdateCb: (valid: boolean, data?: T, errors?: ResolverError<T>) => void
): Promise<ResolverResult<T>> {
  try {
    const values = await schema.validate(data, {
      abortEarly: false,
    });

    onUpdateCb(true);

    return {
      values,
      errors: {},
    };
  } catch (e) {
    onUpdateCb(false);

    return {
      values: {},
      errors: e.inner.reduce(
        (
          allErrors: Record<string, string>,
          currentError: { path: string; type?: string; message: string }
        ) => {
          return {
            ...allErrors,
            [currentError.path]: {
              type: currentError.type ?? 'validation',
              message: currentError.message,
            },
          };
        },
        {}
      ),
    };
  }
}
