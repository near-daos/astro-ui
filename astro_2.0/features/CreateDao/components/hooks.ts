import {
  useFormContext,
  UseFormGetValues,
  UseFormSetValue,
} from 'react-hook-form';
import {
  DAOFormValues,
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
} from 'astro_2.0/features/CreateDao/components/types';
import { useCallback, useEffect, useState } from 'react';
import {
  DAO_PROPOSALS_OPTIONS,
  DAO_STRUCTURE_OPTIONS,
  DAO_VOTING_POWER_OPTIONS,
} from 'astro_2.0/features/CreateDao/components/data';
import { httpService } from 'services/HttpService';
import { CreateDaoCustomInput, CreateDaoInput } from 'types/dao';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import { SputnikWalletError } from 'errors/SputnikWalletError';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';
import {
  getInitialValues,
  updateAction,
} from 'astro_2.0/features/CreateDao/components/helpers';
import { useStateMachine } from 'little-state-machine';
import { useWalletContext } from 'context/WalletContext';
import { GA_EVENTS, sendGAEvent } from 'utils/ga';

type DaoFormStateReturn = {
  options: DaoSettingOption<
    DAOStructureType | DAOVotingPowerType | DAOProposalsType
  >[];
  getValues: UseFormGetValues<DAOFormValues>;
  setValue: UseFormSetValue<DAOFormValues>;
};

export const useDaoFormState = (): DaoFormStateReturn => {
  const { getValues, watch, setValue } = useFormContext<DAOFormValues>();

  const [structure, voting, proposals] = watch([
    'structure',
    'voting',
    'proposals',
  ]);

  const [options, setOptions] = useState<
    DaoSettingOption<DAOStructureType | DAOVotingPowerType | DAOProposalsType>[]
  >([]);

  useEffect(() => {
    setOptions([
      DAO_STRUCTURE_OPTIONS[structure],
      DAO_VOTING_POWER_OPTIONS[voting],
      DAO_PROPOSALS_OPTIONS[proposals],
    ]);
  }, [structure, voting, proposals, setValue]);

  return { options, getValues, setValue };
};

export function useImageUpload(): {
  uploadImage: (file: File) => Promise<string>;
} {
  const uploadImage = useCallback(async (img: File) => {
    try {
      if (img) {
        const { data } = await httpService.post('/api/upload-to-s3', img, {
          baseURL: '',
        });

        return data;
      }
    } catch (e) {
      return '';
    }

    return '';
  }, []);

  return { uploadImage };
}

export function useCreateDao(): {
  uploadAssets: (defaultFlag: string) => Promise<string>;
  createDao: (
    daoName: string,
    data: CreateDaoInput | CreateDaoCustomInput
  ) => Promise<void>;
} {
  const router = useRouter();
  const { t } = useTranslation();
  const { accountId, nearService } = useWalletContext();
  const { actions, state } = useStateMachine({ updateAction });

  const uploadImg = useCallback(async (img: File) => {
    if (img) {
      const { data } = await httpService.post('/api/upload-to-s3', img, {
        baseURL: '',
      });

      return data;
    }

    return '';
  }, []);

  async function loadImage(defaultFlag: string) {
    const response = await fetch(defaultFlag);
    const blob = await response.blob();

    return new File([blob], 'image.png', { type: 'image/png' });
  }

  const uploadAssets = useCallback(
    async (defaultFlag: string) => {
      try {
        const defaultFlagFile = await loadImage(defaultFlag);

        return uploadImg(defaultFlagFile);
      } catch (e) {
        return '';
      }
    },
    [uploadImg]
  );

  const createDao = useCallback(
    async (daoName: string, data: CreateDaoInput | CreateDaoCustomInput) => {
      try {
        await nearService?.createDao(data);

        const { nearConfig } = configService.get();
        const daoId = `${daoName}.${nearConfig?.contractName ?? ''}`;

        sendGAEvent({
          name: GA_EVENTS.CREATE_DAO,
          daoId,
          accountId,
        });

        showNotification({
          type: NOTIFICATION_TYPES.INFO,
          description: t('notifications.transactionDelay'),
          lifetime: 20000,
        });

        sessionStorage.removeItem('__LSM__');
        actions.updateAction(
          getInitialValues(accountId, state.assets.defaultFlag)
        );

        await router.push(`/dao/${daoId}`);
      } catch (error) {
        console.warn(error);

        if (error instanceof SputnikWalletError) {
          showNotification({
            type: NOTIFICATION_TYPES.ERROR,
            description: error.message,
            lifetime: 20000,
          });
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [router, uploadImg, nearService]
  );

  return {
    createDao,
    uploadAssets,
  };
}
