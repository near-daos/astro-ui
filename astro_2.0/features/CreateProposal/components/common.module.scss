@import 'styles/mixins';

// TODO refactor all proposals to use this styles
.input {
  font-weight: 700 !important;

  @include placeholder {
    color: var(--color-neutral-40) !important;
    transition: opacity 0.2s linear !important;
  }

  /* stylelint-disable */
  /* Chrome, Safari, Edge, Opera */
  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  /* Firefox */
  &[type='number'] {
    -moz-appearance: textfield;
  }
  /* stylelint-enable */
}
