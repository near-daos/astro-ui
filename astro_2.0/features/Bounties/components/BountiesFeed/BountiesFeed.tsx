import React, { FC, useCallback, useState } from 'react';
import { useRouter } from 'next/router';
import isEmpty from 'lodash/isEmpty';
import { useAsyncFn, useMountedState } from 'react-use';
import { useTranslation } from 'next-i18next';
import cn from 'classnames';

import { useDebounceEffect } from 'hooks/useDebounceUpdateEffect';
import { useBountySearch } from 'astro_2.0/features/Bounties/components/hooks';

import { BountyContext } from 'types/bounties';
import { DAO } from 'types/dao';
import { PaginationResponse } from 'types/api';

import { Feed as FeedList } from 'astro_2.0/components/Feed';
import { NoResultsView } from 'astro_2.0/components/NoResultsView';
import { SideFilter } from 'astro_2.0/components/SideFilter';
import { Loader } from 'components/loader';
import { ViewBounty } from 'astro_2.0/features/ViewBounty';
import { SearchInput } from 'astro_2.0/components/SearchInput';

import { Tokens } from 'types/token';
import { useWalletContext } from 'context/WalletContext';
import { HideBountyContextProvider } from 'astro_2.0/features/Bounties/components/HideBountyContext';

import { SputnikHttpService } from 'services/sputnik';
import { LIST_LIMIT_DEFAULT } from 'services/sputnik/constants';

import styles from './BountiesFeed.module.scss';

interface BountiesFeedProps {
  initialData: PaginationResponse<BountyContext[]> | null;
  dao?: DAO;
  tokens?: Tokens;
}

const FEED_OPTIONS = [
  {
    label: 'Proposal Phase',
    value: 'proposalPhase',
  },
  {
    label: 'Available Bounty',
    value: 'availableBounty',
  },
  {
    label: 'In Progress',
    value: 'inProgress',
  },
  {
    label: 'Completed',
    value: 'completed',
  },
];

export const BountiesFeed: FC<BountiesFeedProps> = ({ initialData, dao }) => {
  const { accountId } = useWalletContext();
  const { t } = useTranslation();
  const isMounted = useMountedState();
  const { query } = useRouter();
  const [data, setData] = useState(initialData);
  const [loading, setLoading] = useState(false);

  const { handleSearch, loading: searching } = useBountySearch();

  const [{ loading: dataIsLoading }, fetchData] = useAsyncFn(
    async (initData?: typeof data) => {
      let accumulatedListData = initData || null;

      const res = await SputnikHttpService.getBountiesContext(
        query.dao as string,
        accountId,
        {
          offset: accumulatedListData?.data.length || 0,
          limit: LIST_LIMIT_DEFAULT,
          bountySort: query.bountySort as string,
          bountyFilter: query.bountyFilter as string,
          bountyPhase: query.bountyPhase as string,
        }
      );

      if (!res) {
        return null;
      }

      accumulatedListData = {
        ...res,
        data: [...(accumulatedListData?.data || []), ...(res.data || [])],
      };

      // Reset custom loading state
      if (isMounted()) {
        setLoading(false);
      }

      return accumulatedListData;
    },
    [data?.data?.length, query, accountId]
  );

  useDebounceEffect(
    async ({ isInitialCall, depsHaveChanged }) => {
      if (isInitialCall || !depsHaveChanged) {
        return;
      }

      const newProposalsData = await fetchData();

      if (isMounted()) {
        setData(newProposalsData);
      }

      window.scroll(0, 0);
    },
    1000,
    [query.bountyPhase, query.bountySort, query.bountyFilter]
  );

  const loadMore = async () => {
    if (dataIsLoading) {
      return;
    }

    const newProposalsData = await fetchData(data);

    if (isMounted()) {
      setData(newProposalsData);
    }
  };

  const handleBountyInoutSearch = useCallback(
    async val => {
      if (!val || !val.trim()) {
        return null;
      }

      const res = await handleSearch(val);

      if (isMounted()) {
        setData(res);
      }

      return res;
    },
    [handleSearch, isMounted]
  );

  const handleBountyInputClose = useCallback(async () => {
    const newProposalsData = await fetchData();

    if (isMounted()) {
      setData(newProposalsData);
    }
  }, [fetchData, isMounted]);

  return (
    <div className={styles.root}>
      <HideBountyContextProvider>
        <div className={cn(styles.row, styles.additionalFilters)}>
          <SearchInput
            onSubmit={handleBountyInoutSearch}
            loading={searching}
            onClose={handleBountyInputClose}
          />
        </div>
        <div className={cn(styles.row, styles.content)}>
          <SideFilter
            shallowUpdate
            hideAllOption
            queryName="bountyPhase"
            list={FEED_OPTIONS}
            title={t('feed.filters.chooseAFilter')}
            className={styles.filter}
          />
          <div className={styles.container}>
            {loading ? (
              <Loader className={styles.loader} />
            ) : (
              data && (
                <FeedList
                  data={data}
                  loadMore={loadMore}
                  loader={<p className={styles.loading}>{t('loading')}...</p>}
                  noResults={
                    <div className={styles.loading}>
                      <NoResultsView
                        title={
                          isEmpty(data?.data)
                            ? t('noDataFound')
                            : t('noMoreResults')
                        }
                      />
                    </div>
                  }
                  renderItem={bountyContext => (
                    <div
                      key={bountyContext.id}
                      className={styles.bountyCardWrapper}
                    >
                      {dao && bountyContext.proposal && (
                        <ViewBounty
                          contextId={bountyContext.id}
                          commentsCount={bountyContext.commentsCount}
                          daoId={dao.id}
                          bounty={bountyContext.bounty}
                          proposal={bountyContext.proposal}
                          initialInfoPanelView={null}
                        />
                      )}
                    </div>
                  )}
                  className={styles.listWrapper}
                />
              )
            )}
          </div>
        </div>
      </HideBountyContextProvider>
    </div>
  );
};
