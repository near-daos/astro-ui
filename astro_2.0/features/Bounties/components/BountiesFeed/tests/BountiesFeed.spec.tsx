import React from 'react';
import { render } from 'jest/testUtils';

import { BountiesFeed } from 'astro_2.0/features/Bounties/components/BountiesFeed';

describe('BountiesFeed', () => {
  it('Should render component', () => {
    const data = {
      count: 0,
      total: 0,
      page: 0,
      pageCount: 0,
      data: [],
    };

    const { container } = render(<BountiesFeed initialData={data} />);

    expect(container).toMatchSnapshot();
  });
});
