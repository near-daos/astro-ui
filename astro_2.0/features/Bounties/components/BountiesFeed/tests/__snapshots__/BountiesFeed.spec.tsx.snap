// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BountiesFeed Should render component 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="row additionalFilters"
    >
      <div
        class="root"
      >
        <div
          class="container"
          role="button"
          tabindex="0"
        >
          <div
            class="iconHolder"
          >
            <svg
              class="icon"
            >
              <use
                href="/_next/undefined"
              />
            </svg>
          </div>
          <input
            class="input"
            placeholder="searchBountyPlaceholder"
            tabindex="0"
            type="text"
            value=""
          />
          <div
            class="anchor"
          />
        </div>
      </div>
    </div>
    <div
      class="row content"
    >
      <div
        class="categoriesListRoot filter"
      >
        <p
          class="categoriesListTitle"
        >
          feed.filters.chooseAFilter
        </p>
        <ul
          class="categoriesList"
          style="--filter-active-transform: 0%;"
        >
          <li
            class="categoriesListItemWrapper"
          >
            <a
              class="categoriesListItem"
              href="/?bountyPhase=proposalPhase"
              tabindex="0"
            >
              Proposal Phase
            </a>
          </li>
          <li
            class="categoriesListItemWrapper"
          >
            <a
              class="categoriesListItem"
              href="/?bountyPhase=availableBounty"
              tabindex="0"
            >
              Available Bounty
            </a>
          </li>
          <li
            class="categoriesListItemWrapper"
          >
            <a
              class="categoriesListItem"
              href="/?bountyPhase=inProgress"
              tabindex="0"
            >
              In Progress
            </a>
          </li>
          <li
            class="categoriesListItemWrapper"
          >
            <a
              class="categoriesListItem"
              href="/?bountyPhase=completed"
              tabindex="0"
            >
              Completed
            </a>
          </li>
        </ul>
      </div>
      <div
        class="container"
      >
        <div
          class="root listWrapper"
        >
          <div
            class="infinite-scroll-component__outerdiv"
          >
            <div
              class="infinite-scroll-component "
              style="height: auto; overflow: initial;"
            >
              <div
                class="loading"
              >
                <div
                  class="root"
                >
                  <div
                    class="image"
                  />
                  <div
                    class="title"
                  >
                    noDataFound
                  </div>
                  <div
                    class="desc"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="selection"
          />
        </div>
      </div>
    </div>
    <div
      class="hideControlButtonWrapper"
    />
  </div>
</div>
`;
